Editor GUI Fixes

rogerbarton opened the pull request:
1. Editors (component/window) are repainted when dll's are un/loaded using the new attributes
   - Involves running the triggers on the main thread in a queue 39c4fce825d7f1c4572d183ec8666593d3840f13
1. Editor variable changes are detected and saved properly
   - Involves creating a clone of the options and comparing changes when the GUI has changed 5bda99031c315006d66ea14aa8c7c949c7dabd6e

Note: The attributes `[NativeDllLoadedTrigger]` etc are not searched for in the source code of this tool itself. So by creating an `.asmdef` in #14 these attributes may not work, do you maybe know of an easy fix for this @mcpiroman?.

rogerbarton committed:
Repaint editor GUIs on shortcut

rogerbarton committed:
Editor variable changes saved properly

Before changing a variable would not be detected and so not saved/serialized in some cases.

rogerbarton committed:
Use callback attributes to repaint editors

This would not work with the `upm-support` branch #14 as the attributes are not searched for in the attribute that these scripts are in by default.

mcpiroman commented on scripts/Editor/DllManipulatorWindowEditor.cs:
```diff
@@ -6,13 +6,22 @@ namespace UnityNativeTool.Internal
+        [NativeDllAfterUnloadTrigger]
+        public static void RepaintAll()
+        {
+            if(window) window.Repaint();
```
Out of curiosity, repainting an editor inside a window doesn't also make the window repaint?

And as a nit, I don't use one line control blocks in this project (and anywhere else). So just an enter here.

rogerbarton replied:
Yes I tested it today with the new changes 74c1c96f19b432179a9780be3f3e0ac42a71aff8 and you seem to have to also repaint the window.

>So just an enter here

It's no longer in there but sure.

mcpiroman commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -8,6 +8,8 @@
 using System.IO;
 using System;
+using UnityEditor.SceneManagement;
+using UnityEngine.SceneManagement;
```
Duplicated?

rogerbarton replied:
No we need it both from the Editor and Engine. They are not the same.

rogerbarton replied:
I realized we dont actually need to set the scene as dirty when we set the gui target as dirty. So its been removed. 87dcb2ab6fde2a1a1712ec37ea1277a1b2df138f

mcpiroman commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -333,5 +341,15 @@ public static void UnloadAll()
+        public static void RepaintAll()
+        {
+            var editors = Resources.FindObjectsOfTypeAll<DllManipulatorEditor>();
+            if(editors == null) return;
```
nit: format, see below

mcpiroman reviewed (commented):
> Note: The attributes [NativeDllLoadedTrigger] etc are not searched for in the source code of this tool itself

Yeah, I just search for them in the same assemblies as for native function declarations. I guess I should just search every assembly and maybe exclude standard unity ones for perf, because searching in every assembly doesn't sound the quickest thing.

rogerbarton committed:
Allow multiple un/loadTrigger attributes, use action to trigger repaints

rogerbarton committed:
Remove SceneManagement, it's not required

The scene is automatically set as dirty when the gui target is

rogerbarton commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -139,6 +142,11 @@ public override void OnInspectorGUI()
                 var time = t.InitializationTime.Value;
                 EditorGUILayout.LabelField($"Initialized in: {(int)time.TotalSeconds}.{time.Milliseconds.ToString("D3")}s");
             }
+
+            if (GUI.changed)
+            {
+                EditorUtility.SetDirty(target);
+            }
```
I've realized that this will set the scene as dirty whenever the GUI changes, even if the serialized data doesn't change. So it will report that the scene needs to be saved when there are no changes. Any ideas how to detect this in a better way than `GUI.changed`? 

This is still better than previously where we lost changes sometimes.

mcpiroman replied:
Sorry but I don't know of any better way. Looks like something which should be trivial enough, but very well is not.

rogerbarton replied:
It appears to only be the case when the foldouts/buttons are clicked. So dropdown's, tooltips don't set `GUI.changed`.

rogerbarton replied:
The best way I see is to have a copy of the previous options and compare them if `GUI.changed` to check for changes. However, this seems like a bit of a overkill and requires various stuff with cloning and `Equals`. For now I would leave it like above.

rogerbarton replied:
Implemented in 5bda99031c315006d66ea14aa8c7c949c7dabd6e

rogerbarton committed:
Merge branch 'master' into shortcuts

rogerbarton committed:
Fixes to gui with enableInEdit mode

Admittedly this should have been in #12. However, here are the fixes anyways.

GUI buttons to un/load show in edit mode if enableInEditMode is true. Disabling and re-enabling the DllManipulatorScript works now.

rogerbarton committed:
Properly reset custom triggers (fix duplicates)

Previously triggers would not be cleared properly and so be duplicated. (Static variables seem to persist between entering/exiting playmode)

rogerbarton committed:
Custom triggers optional execute on main thread

Allow custom triggers to be executed on the main thread in a queue. We need this internally to repaint the editor GUI as it uses the Unity API.

rogerbarton committed:
Update README.md

rogerbarton committed:
Mark scene dirty only if options changed

rogerbarton commented on scripts/DllManipulatorScript.cs:
```diff
@@ -46,7 +46,7 @@ private void OnEnable()
                 if (EditorApplication.isPlaying)
                     Destroy(gameObject);
-                else
+                else if(_singletonInstance != this)
```
Allows the user to properly toggle enabled in edit mode.

rogerbarton commented:
Ready for review @mcpiroman.

You will probably want to review 39c4fce825d7f1c4572d183ec8666593d3840f13 separately.

(side note: #14 should already be compatible with these changes)

rogerbarton committed:
Bug fix, resetting when not initialized

Initialize() is not always called so we should not always reset.

rogerbarton committed:
small fix

mcpiroman commented on scripts/Attributes.cs:
```diff
@@ -30,13 +30,24 @@ public class DisableMockingAttribute : Attribute
+        /// <summary>
+        /// Should the method always be executed on the main thread, to allow use of the Unity API.
+        /// Note: this means the method is not immediately executed but put in a queue.
+        /// For consistent behaviour, the method is put in the queue even if it is triggered from the main thread.
```
Just about the line 39 - I don't think this behavior is really beneficial, I imagine usually one would want main thread events to stay in order. And as of consistency, dispatching frameworks which I recall tend to invoke calls from target ("main") thread right away. For events though, while it doesn't prevent dead-locks as much as with direct calls, I think this also makes sense.

rogerbarton replied:
Ok I've changed it. I wasn't quite sure either which is why I left it commented out:
https://github.com/rogerbarton/UnityNativeTool/blob/a0dc185babf73065dad5935be87e2de2f8d5811a/scripts/DllManipulator.cs#L547-L548

mcpiroman commented on scripts/DllManipulator.cs:
```diff
@@ -509,20 +526,41 @@ internal static void LoadTargetFunction(NativeFunction nativeFunction, bool igno
+                
+                // Execute now or queue to the main thread
+                if (useMainThreadQueue /*&& Thread.CurrentThread.ManagedThreadId != _unityMainThreadId*/)
+                    _mainThreadTriggerQueue.Enqueue(new Tuple<MethodInfo, object[]>(methodInfo, args));
```
Instead of using tuple to wrap arguments, it could be simpler to just store an `Action` created with lambda, like so:
`_mainThreadTriggerQueue.Enqueue(() => methodInfo.Invoke(null, args));` (You could also make e.g. `Action<int>` to provide an argument at call side, JIC you didn't know).

rogerbarton replied:
Yes makes sense

mcpiroman commented on scripts/DllManipulator.cs:
```diff
@@ -101,18 +103,33 @@ internal static void Initialize(int unityMainThreadId, string assetsPath)
             else
             {
-                Debug.LogError($"Trigger method must either take no parameters or one parameter of type {nameof(NativeDll)}. Violation on method {method.Name} in {method.DeclaringType.FullName}");
+                Debug.LogError($"Trigger method must either take no parameters or one parameter of type {nameof(NativeDll)} or two of type {nameof(NativeDll)} and int. Violation on method {method.Name} in {method.DeclaringType.FullName}");
```
`or two of type {nameof(NativeDll)} and int` at first reading this sounded like two NativeDll parameters and one int parameter, don't know if that's just me though.
Also why do you assume the main thread id information would be useful? And when it turns out to be, it should be documented somehow (or I just missed that, in which case I'm sorry)

rogerbarton replied:
I've changed it to be clearer and given examples of the possible functions in the `TriggerAttribute` doc.

The `mainThreadId` can be used to tell if we are on the main thread and are allowed to call unity functions. I guess this is partially obsolete now that we have the `MainThreadTriggerQueue`.

mcpiroman commented on scripts/DllManipulator.cs:
```diff
@@ -509,20 +526,41 @@ internal static void LoadTargetFunction(NativeFunction nativeFunction, bool igno
+        /// Executes queued methods.
+        /// Should be called from the main thread in Update.
+        /// </summary>
+        public static void InvokeMainThreadQueue()
```
Don't wanna be fussy, but I'd move the dispatcher stuff to `DllManipulatorScript`, as it is closer to Unity, whilst this is intended to do 'the stuff', if you will.

rogerbarton replied:
I assume you only mean to move this function, not the queue itself

mcpiroman replied:
I mean the queue itself. I don't quite imagine how would you move just this method, so that might be interesting too.

mcpiroman replied:
I mean the queue itself. I don't quite imagine how you would move just the method, so I don't disagree with that now.

rogerbarton replied:
By making the queue public (its static already)

mcpiroman replied:
Ah yes, but no, I see this whole in the Script as it is the interface with Unity.

mcpiroman commented on scripts/DllManipulatorScript.cs:
```diff
@@ -57,6 +57,10 @@ private void OnEnable()
                 Initialize();
+            
+            if(!EditorApplication.isPlaying && Options.enableInEditMode)
+                EditorApplication.update += Update;
```
Doens't just having `[ExecuteInEditMode]` make the `Update` be invoke in editor? Or do I misunderstand sth.

rogerbarton replied:
For `ExecuteInEditMode` "Update is only called when something in the Scene changed" - [Docs](https://docs.unity3d.com/ScriptReference/ExecuteInEditMode.html). So we need to manually add this to cover all cases.

mcpiroman commented on scripts/Editor/DllManipulatorEditor.cs:
```diff
@@ -339,5 +362,12 @@ public static void UnloadAll()
+        [NativeDllAfterUnloadTrigger(UseMainThreadQueue = true)]
+        public static void RepaintAll()
+        {
+            RepaintAllEditors.Invoke();
```
It's usually safer to handle null case with events (which can be done simply by putting ? before the dot). In this case you initialize the event, however I'm not sure if adding and removing an listener doesn't reset it back to null. Either or, it's better to be safe.

rogerbarton replied:
Agreed however as the event is initialized with the default delegate it will never be null. https://stackoverflow.com/q/170907/9295437
Ill add it anyways...

mcpiroman reviewed (commented):
Sorry this took me this long to have a look at this.
I'm somewhat scared about that option cloning, comparing (mainly bout forgetting to update that), but (partly because I'm not quite in the problem it tries to solve), I don't see a better way.

rogerbarton committed:
Review fixes

Also RegisterTriggerMethod receives attribute instead of bool, 
Fix naming in DllManipulatorOptions

rogerbarton commented:
Regarding cloning the Options, its just that we need a separate copy for detecting changes. This copy is never changed directly only update when changes are made. Its pretty safe in my opinion. 

I've actually found this quite a good solution.

rogerbarton committed:
Add comments

mcpiroman commented:
I'm fine with that, thanks!
